src/frontend/src/optimizer/rule/pull_up_correlated_predicate_agg_rule.rs

```diff
@@ -145,12 +157,52 @@ impl Rule for PullUpCorrelatedPredicateAggRule {
 
         // If there is a count aggregate, bail out and leave for general subquery unnesting to deal.
         // When group by is empty, count would return 0 instead of null.
-        if agg_calls
+        // Unless we can prove that the row generated by the empty group could be eliminated by the top filter.
+        //
+        // A typical example is the count is generate by avg, because avg = sum / count. In this case, when input is empty,
+        // sum is null, so avg is null. And null-rejected expression will be false, so we can still apply this rule and we don't need to generate a 0 value for count.
+        let count_exists = agg_calls
             .iter()
-            .any(|agg_call| agg_call.agg_kind == AggKind::Count)
-        {
-            return None;
-        };
+            .any(|agg_call| agg_call.agg_kind == AggKind::Count);
+
+        if count_exists {
+            // When group input is empty, not count agg would return null.
+            let null_agg_pos = agg_calls
+                .iter()
+                .positions(|agg_call| agg_call.agg_kind != AggKind::Count)
+                .collect_vec();
+
+            // If no null agg, bail out.
+            if null_agg_pos.is_empty() {
+                return None;
+            }
+
+            // Try to prove that the expression is null-rejected by the top filter when not-count-aggs are null.
+            let mut agg_null_bitset = FixedBitSet::with_capacity(agg.base.schema().len());
+            for pos in null_agg_pos {
+                agg_null_bitset.insert(pos);
+            }
```

Comment:
Are you looking for [`FixedBitSet::from_iter`](https://docs.rs/fixedbitset/0.4.2/fixedbitset/struct.FixedBitSet.html#impl-FromIterator%3Cusize%3E)?

Reply:
I need to ensure the capacity of this bitset is `agg.base.schema().len()`.